### ISSUE TITLE
The broadcast transmission frequency is low

### ISSUE THREAD

**Issue by shangsony** [ljedrz/pea2pea#9]
Test code:
```
use bytes::Bytes;
use once_cell::sync::Lazy;
use rand::{distributions::Standard, rngs::SmallRng, Rng, SeedableRng};
use tokio::time::sleep;
use tracing::*;
mod common;
use pea2pea::{
    protocols::{Reading, Writing},
    Config, Node, Pea2Pea,
};
use std::time::Duration;

use std::{convert::TryInto, io, net::SocketAddr, time::Instant};

const NUM_MESSAGES: usize = 10_000;
const MSG_SIZE: usize = 200;

static RANDOM_BYTES: Lazy<Bytes> = Lazy::new(|| {
    Bytes::from(
        (&mut SmallRng::from_entropy())
            .sample_iter(Standard)
            .take(MSG_SIZE - 2)
            .collect::<Vec<_>>(),
    )
});

#[derive(Clone)]
pub struct P2PNode(pub Node);

impl Pea2Pea for P2PNode {
    fn node(&self) -> &Node {
        &self.0
    }
}

#[async_trait::async_trait]
impl Reading for P2PNode {
    type Message = bytes::Bytes;

    fn read_message<R: io::Read>(
        &self,
        _source: SocketAddr,
        reader: &mut R,
    ) -> io::Result<Option<Self::Message>> {
        let vec = crate::common::read_len_prefixed_message::<R, 2>(reader)?;
        Ok(vec.map(bytes::Bytes::from))
    }

    async fn process_message(&self, _src: SocketAddr, _msg: Self::Message) -> io::Result<()> {
        Ok(())
    }
}

impl Writing for P2PNode {
    type Message = bytes::Bytes;

    fn write_message<W: io::Write>(
        &self,
        _target: SocketAddr,
        payload: &Self::Message,
        writer: &mut W,
    ) -> io::Result<()> {
        writer.write_all(&(payload.len() as u16).to_le_bytes())?;
        writer.write_all(payload)
    }
}

impl P2PNode {
    pub async fn new<T: Into<String>>(name: T) -> Self {
        let config = Config {
            name: Some(name.into()),
            ..Default::default()
        };
        Self(Node::new(Some(config)).await.unwrap())
    }

    fn broadcast(&self) -> io::Result<()>{
    for addr in self.node().connected_addrs() {
        let self_clone = self.clone();
        tokio::spawn(async move {
            self_clone.send_direct_message(addr,RANDOM_BYTES.clone());       
            });
        }
    Ok(())
}


    fn send_periodic_broadcasts(&self) {
        let self_clone = self.clone();
        tokio::spawn(async move {
            loop {
                if self_clone.node().num_connected() != 0 {
                    self_clone.send_broadcast(RANDOM_BYTES.clone()).unwrap();
                    //self_clone.broadcast();
                } else {
                    info!(parent: self_clone.node().span(), "meh, I have no frens to chat with",);
                }
                sleep(Duration::from_micros(1)).await;
            }
        });
    }
}

#[tokio::test]
async fn broadcast_benches_example() {
    // tracing_subscriber::fmt::init();
    // println!("{}", RANDOM_BYTES.len());

    let random_nodes = common::start_nodes(4, None)
        .await
        .into_iter()
        .map(P2PNode)
        .collect::<Vec<_>>();
    for rando in &random_nodes {
        rando.enable_reading().await;
    }

    let broadcaster = P2PNode::new("chatty").await;

    broadcaster.enable_writing().await;
    broadcaster.send_periodic_broadcasts();

    for rando in &random_nodes {
        broadcaster
            .0
            .connect(rando.node().listening_addr().unwrap())
            .await
            .unwrap();
    }

    let now = std::time::Instant::now();
    loop {
        tokio::time::sleep(std::time::Duration::from_millis(1)).await;
        if now.elapsed() > std::time::Duration::from_secs(60) {
            break;
        }
    }

    let r = broadcaster.node().stats().sent();
    println!("broadcaster send msgs({}) bytes({})", r.0, r.1);
    for rando in &random_nodes {
        let r = rando.node().stats().received();
        println!("receiver receive msgs({}) bytes({})", r.0, r.1);
    }
}
```
result：
![image](https://user-images.githubusercontent.com/19890858/154417385-a3620c6c-e3cd-4575-9138-5a565e0f0657.png)

suggest：
Can you optimize the low performance of sending broadcast processing

**Comment by ljedrz** [comment]
There are a few things to tackle here:
- my results are way better than yours right away, which might suggest you have few cores on your PC; my Ryzen 9 3900X gave me the following:
```
broadcaster send msgs(224760) bytes(44952000)
receiver receive msgs(56190) bytes(11238000)
receiver receive msgs(56190) bytes(11238000)
receiver receive msgs(56190) bytes(11238000)
receiver receive msgs(56190) bytes(11238000)
```
- the broadcast interval you set is WAY too low; I changed it to 100us (which I still consider overkill for any purpose) and saw little change
- the message size is quite low - you are unlikely to get good throughput with any implementation when you're sending messages this small; in the existing benchmark for a single reader with multiple writers the message size that gave the best results was `32KiB`
- all this message sending and processing is happening within the same async runtime, which means the more messages are sent, the worse performance you will observe _within the same machine_; you are not just measuring broadcast performance here, you're getting a performance penalty for all the reading as well
- the inbound queues of the receiver nodes are set to the default value of `64`, which means they are likely discarding many of the messages they receive (due to their inbound queue being full), leading to lower results

Summing up, I don't believe there are any inherent issues with the broadcast implementation. It is very similar to the one for direct messages, which can reach a throughput of >10GB/s in the existing `bench_spam_to_one` benchmark. There is a small performance penalty due to the use of boxed trait objects, but it's unlikely to matter in normal network conditions, where the network itself is the slowest bit.

It's also _really_ hard to have good, informative benchmarks for a library like this, and I remember attempting to have a benchmark similar to yours, and in the end deciding that it was not practical enough.

Nevertheless, I'm happy to suggest some improvements if you can tell me what your specific use-case is.

**Comment by shangsony** [comment]
hi,I used the broadcast function to send the broadcast frequently, the size of the broadcast is about 50KB, and the frequent broadcast is sent. I made the statistics of the time from the broadcast node to other receiver nodes, but the time is about 10 seconds, please optimize  
![image](https://user-images.githubusercontent.com/19890858/154937636-215af3a2-892f-4c74-aebd-a001cf52c701.png)

**Comment by ljedrz** [comment]
How costly is your serialization? As [the docs](https://docs.rs/pea2pea/0.33.0/pea2pea/protocols/trait.Writing.html#associatedtype.Message) state:
> unless their serialization is expensive and the message is broadcasted (in which case it would get serialized multiple times), serialization should be done in [Writing::write_message](https://docs.rs/pea2pea/0.33.0/pea2pea/protocols/trait.Writing.html#tymethod.write_message)

In other words, have you considered using `Bytes` or `Arc<[u8]>` as the type for `Writing::Message`, which would allow serialization to be done only once (`Writing::write_message` would then not need to serialize the message anymore)?

**Comment by shangsony** [comment]
The way I test it is with 4 nodes in different processes rather than 4 nodes in the same process

**Comment by shangsony** [comment]
I using Bytes or Arc<[u8]> as the type for Writing::Message,But it didn't get any better, so it wasn't serialization
The way I test it is with 4 nodes in different processes rather than 4 nodes in the same process;
it is my code:
#[async_trait::async_trait]
impl Reading for PeerNetwork {
    type Message = bytes::Bytes;
    fn read_message<R: io::Read>(
        &self,
        source: SocketAddr,
        reader: &mut R,
    ) -> io::Result<Option<Self::Message>> {
        let vec = P2PUtils::read_len_prefixed_message::<R, 4>(reader)?;
        Ok(vec.map(bytes::Bytes::from))
    }

    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
        if let Some(handle) = self.reading_buffer_handler(){
           handle.send((source,message));
        }
        Ok(())
    }
}

impl Writing for PeerNetwork {
    type Message = bytes::Bytes;

    fn write_message<W: io::Write>(
        &self,
        _: SocketAddr,
        payload: &Self::Message,
        writer: &mut W,
    ) -> io::Result<()> {
        writer.write_all(&(payload.len() as u32).to_le_bytes())?;
        writer.write_all(&payload)
    }
}

**Comment by ljedrz** [comment]
> The way I test it is with 4 nodes in different processes rather than 4 nodes in the same process

This is all still subject to a single `tokio` executor, so just adding processes (threads or tasks) won't help you at all; have you checked if the performance changes more or less linearly with an increase/decrease in the number of nodes?

**Comment by shangsony** [comment]
Hi, I know what the cause is, because the broadcast message
( fn send_broadcast(&self, message: Self::Message) -> io::Result<()>) 
and direct message
(fn send_direct_message(
        &self,
        addr: SocketAddr,
        message: Self::Message,
    ) -> io::Result<oneshot::Receiver<bool>>) 
are used in different threads at the same time, can you internally implement the relationship between the two interfaces to prevent resource competition  ,After I remove the direct message, and then send a broadcast message program to the normal broadcast

**Comment by shangsony** [comment]
Hi, I know what the cause is, because the broadcast message
( fn send_broadcast(&self, message: Self::Message) -> io::Result<()>) 
and direct message
(fn send_direct_message(
        &self,
        addr: SocketAddr,
        message: Self::Message,
    ) -> io::Result<oneshot::Receiver<bool>>) 
are used in different threads at the same time, can you internally implement the relationship between the two interfaces to prevent resource competition  ,After I remove the direct message, and then send a broadcast message program to the normal broadcast

**Comment by ljedrz** [comment]
As I mentioned before, both methods use the same `tokio` executor internally, and there is no reason to have different executors for them. These methods barely do anything other than sending the message to the task dedicated to writing messages.

In any case, any resource exhaustion related to running many nodes locally is not an actual issue you'd experience running with a proper network due to network delays (as opposed to having to read messages immediately once they are written in local conditions).

**Comment by shangsony** [comment]
In p2p network, one node is deployed separately and the configuration information of other nodes is provided. It is not possible to run several nodes in a process randomly as you do

**Comment by ljedrz** [comment]
This is likely a language barrier, but I don't understand the nature of your issue.

**Comment by shangsony** [comment]
Hello! In a normal P2P network, the connection or network shape of the P2P network must be independently deployed by each node, and each node is connected to other nodes

**Comment by ljedrz** [comment]
Well, yes; how does this relate in any way to the perceived issue of slow broadcasts?

**Comment by shangsony** [comment]
Like this interface
{ fn send_direct_message(
        &self,
        addr: SocketAddr,
        message: Self::Message,
    ) -> io::Result<oneshot::Receiver<bool>> {
} does not need to wait for the actual arrival, can we remove await block?

**Comment by ljedrz** [comment]
This is [the source](https://docs.rs/pea2pea/latest/src/pea2pea/protocols/writing.rs.html#157-178) for that method:
```
fn send_direct_message(
    &self,
    addr: SocketAddr,
    message: Self::Message,
) -> io::Result<oneshot::Receiver<bool>> {
    // access the protocol handler
    if let Some(handler) = self.node().protocols.writing_handler.get() {
        // find the message sender for the given address
        if let Some(sender) = handler.senders.read().get(&addr).cloned() {
            let (msg, delivery) = WrappedMessage::new(Box::new(message));
            sender.try_send(msg).map_err(|e| {
                error!(parent: self.node().span(), "can't send a message to {}: {}", addr, e);
                self.node().stats().register_failure();
                io::ErrorKind::Other.into()
            }).map(|_| delivery)
        } else {
            Err(io::ErrorKind::NotConnected.into())
        }
    } else {
        Err(io::ErrorKind::Unsupported.into())
    }
}
```
there are no awaits here at all; neither are there any in the default implementation of `Writing::send_broadcast`.

**Comment by shangsony** [comment]
Like this interface does not need to wait for the actual arrival, can we remove await block?  
Through the test to send the broadcast interval set to 50 milliseconds, the broadcast speed is very fast, when the broadcast interval is less than 50 milliseconds will appear message blocking phenomenon

**Comment by shangsony** [comment]
Through the test to send the broadcast interval set to 50 milliseconds, the broadcast speed is very fast, when the broadcast interval is less than 50 milliseconds will appear message blocking phenomenon

**Comment by ljedrz** [comment]
> when the broadcast interval is less than 50 milliseconds will appear message blocking phenomenon

Have you observed it anywhere else than locally? I've mentioned that with many local nodes - especially readers - all sharing the same async executor, they will compete for resources, which is perfectly expected.

**Comment by shangsony** [comment]
I'm sending a lot of messages, and the number of messages that I send is extremely unstable if it goes above 10,000, and then it gets longer and longer

**Comment by shangsony** [comment]
The message size is 100 byte  x 100 =10KB

**Comment by ljedrz** [comment]
But locally or between nodes with their own resources? If locally, then it's expected.

**Comment by shangsony** [comment]
locally

**Comment by shangsony** [comment]
my resources CPU 32 core ，mem 64GB，so resources are not the problem

**Comment by shangsony** [comment]
I can email you my stress test program

**Comment by ljedrz** [comment]
Please do - or post it here - in the most compact version that reproduces the issue.

**Comment by shangsony** [comment]
This cannot commit the zip package

**Comment by shangsony** [comment]
[1.zip](https://github.com/ljedrz/pea2pea/files/8160309/1.zip)
this is my code

**Comment by ljedrz** [comment]
It's not a minified case at all, and there's plenty that could make things considerably slower in your impl: you're hashing inbound messages with a cryptographically secure hash, you're deduplicating them, etc.

**Comment by shangsony** [comment]
Sending messages and hashing are in different threads

**Comment by shangsony** [comment]
Broadcast message processing time, a large part of the time is consumed on the network

**Comment by ljedrz** [comment]
Closing due to inactivity; I haven't been able to reproduce this, unless `Writing::Message::clone` (or `Writing::Codec::encode`) is an expensive operation (in which case it's outside of the scope of `Writing::send_broadcast`).
